# Sorting and cleaning 
---



In order to effectively analyse a dataset, often we need to prepare it first. 
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g. 
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make 
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted* 

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score 
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [1]:
import pandas as pd
import numpy as np
url= "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
happiness_data = pd.read_excel(url)
sorted_data = happiness_data.sort_values("Happiness Score")
df1 = sorted_data.head(5)
display (df1)

    

Country                           Region  Happiness Rank  \
157     Togo               Sub-Saharan Africa             158   
156  Burundi               Sub-Saharan Africa             157   
155    Syria  Middle East and Northern Africa             156   
154    Benin               Sub-Saharan Africa             155   
153   Rwanda               Sub-Saharan Africa             154   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
157            2.839         0.06727                   0.20868  0.13995   
156            2.905         0.08658                   0.01530  0.41587   
155            3.006         0.05015                   0.66320  0.47489   
154            3.340         0.03656                   0.28665  0.35386   
153            3.465         0.03464                   0.22208  0.77370   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
157                   0.28443  0.36453                        0.10731   
156                   0.22396  0.11850                        0.10062   
155                   0.72193  0.15684                        0.18906   
154                   0.31910  0.48450                        0.08010   
153                   0.42864  0.59201                        0.55191   

     Generosity  Dystopia Residual  
157     0.16681            1.56726  
156     0.19727            1.83302  
155     0.47179            0.32858  
154     0.18260            1.63328  
153     0.22628            0.67042

### Exercise 2 - sort by multiple columns, display the first 5 rows 
---

1. sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order 
2. display the first 5 rows of sorted data 

**Test output**:  
Records 122, 127, 147, 100, 96

In [12]:
url= "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
happiness_data = pd.read_excel(url)
sorted = happiness_data.sort_values(by = ["Economy (GDP per Capita)", "Health (Life Expectancy)"])
df2 = sorted.head(5)
display (df2)



Country              Region  Happiness Rank  Happiness Score  \
119  Congo (Kinshasa)  Sub-Saharan Africa             120            4.517   
156           Burundi  Sub-Saharan Africa             157            2.905   
130            Malawi  Sub-Saharan Africa             131            4.292   
143             Niger  Sub-Saharan Africa             144            3.845   
115           Liberia  Sub-Saharan Africa             116            4.571   

     Standard Error  Economy (GDP per Capita)   Family  \
119         0.03680                   0.00000  1.00120   
156         0.08658                   0.01530  0.41587   
130         0.06130                   0.01604  0.41134   
143         0.03602                   0.06940  0.77265   
115         0.11068                   0.07120  0.78968   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
119                   0.09806  0.22605                        0.07625   
156                   0.22396  0.11850                        0.10062   
130                   0.22562  0.43054                        0.06977   
143                   0.29707  0.47692                        0.15639   
115                   0.34201  0.28531                        0.06232   

     Generosity  Dystopia Residual  
119     0.24834            2.86712  
156     0.19727            1.83302  
130     0.33128            2.80791  
143     0.19387            1.87877  
115     0.24362            2.77729

### Exercise 3 - sorting in descending order 
---
 
Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [20]:
url= "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
happiness_data = pd.read_excel(url)

sort_freedom = happiness_data.sort_values(by = ["Freedom", "Trust (Government Corruption)"], ascending = False)
df3 = sort_freedom.tail(5) #

display (df3)


Country                           Region  Happiness Rank  \
136                  Angola               Sub-Saharan Africa             137   
117                   Sudan               Sub-Saharan Africa             118   
95   Bosnia and Herzegovina       Central and Eastern Europe              96   
101                  Greece                   Western Europe             102   
111                    Iraq  Middle East and Northern Africa             112   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
136            4.033         0.04758                   0.75778  0.86040   
117            4.550         0.06740                   0.52107  1.01404   
95             4.949         0.06913                   0.83223  0.91916   
101            4.857         0.05062                   1.15406  0.92933   
111            4.677         0.05232                   0.98549  0.81889   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
136                   0.16683  0.10384                        0.07122   
117                   0.36878  0.10081                        0.14660   
95                    0.79081  0.09245                        0.00227   
101                   0.88213  0.07699                        0.01397   
111                   0.60237  0.00000                        0.13788   

     Generosity  Dystopia Residual  
136     0.12344            1.94939  
117     0.19062            2.20857  
95      0.24808            2.06367  
101     0.00000            1.80101  
111     0.17922            1.95335

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values 

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values 
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [59]:
url= "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
london_data = pd.read_csv(url)
print(london_data.isnull().values.any())

london_data.info()






True
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 5 - remove null values 
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [27]:
url= "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
london_data = pd.read_csv(url)

london_data1 = london_data.dropna(subset = ["life_satisfaction"])
print (london_data1)
#removing all NA values across whole dataframe
display(london_data1.dropna())






           code                  area        date  median_salary  \
613   E09000002  barking and dagenham  2011-12-01        28201.0   
614   E09000003                barnet  2011-12-01        30237.0   
615   E09000004                bexley  2011-12-01        28638.0   
616   E09000005                 brent  2011-12-01        26772.0   
617   E09000006               bromley  2011-12-01        28163.0   
...         ...                   ...         ...            ...   
1010  E12000009            south west  2018-12-01        27956.0   
1013  E92000001               england  2018-12-01        29856.0   
1017  N92000002      northern ireland  2018-12-01        27101.0   
1018  S92000003              scotland  2018-12-01        29289.0   
1019  W92000004                 wales  2018-12-01        26353.0   

      life_satisfaction mean_salary recycling_pct  population_size  \
613                7.05       33568            30         187029.0   
614                7.43       33062        

code                  area        date  median_salary  \
613   E09000002  barking and dagenham  2011-12-01        28201.0   
614   E09000003                barnet  2011-12-01        30237.0   
615   E09000004                bexley  2011-12-01        28638.0   
616   E09000005                 brent  2011-12-01        26772.0   
617   E09000006               bromley  2011-12-01        28163.0   
...         ...                   ...         ...            ...   
999   E09000031        waltham forest  2018-12-01        30298.0   
1000  E09000032            wandsworth  2018-12-01        34501.0   
1001  E09000033           westminster  2018-12-01        43015.0   
1008  E12000007                london  2018-12-01        38146.0   
1013  E92000001               england  2018-12-01        29856.0   

      life_satisfaction mean_salary recycling_pct  population_size  \
613                7.05       33568            30         187029.0   
614                7.43       33062            34         357538.0   
615                7.42       31812            54         232774.0   
616                7.11       29609            37         312245.0   
617                7.50       32863            50         310554.0   
...                 ...         ...           ...              ...   
999                7.46       32875            32         276700.0   
1000               7.65       45317            23         326474.0   
1001               7.66       63792            22         255324.0   
1008               7.58       52629            33        8908081.0   
1013               7.71       37313            44       55977178.0   

      number_of_jobs   area_size  no_of_houses  borough_flag  
613          54000.0      3780.0       71079.0             1  
614         147000.0      8675.0      139346.0             1  
615          78000.0      6429.0       95037.0             1  
616         115000.0      4323.0      112083.0             1  
617         119000.0     15013.0      135036.0             1  
...              ...         ...           ...           ...  
999          88000.0      3881.0      103029.0             1  
1000        147000.0      3522.0      146162.0             1  
1001        775000.0      2203.0      124509.0             1  
1008       6148000.0    159471.0     3556161.0             0  
1013      30493000.0  13303728.0    24172166.0             0  

[267 rows x 12 columns]

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [28]:
url= "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
london_data = pd.read_csv(url)

london_data.drop_duplicates(subset = ['area'])


code                      area        date  median_salary  \
0   E09000001            city of london  1999-12-01        33020.0   
1   E09000002      barking and dagenham  1999-12-01        21480.0   
2   E09000003                    barnet  1999-12-01        19568.0   
3   E09000004                    bexley  1999-12-01        18621.0   
4   E09000005                     brent  1999-12-01        18532.0   
5   E09000006                   bromley  1999-12-01        16720.0   
6   E09000007                    camden  1999-12-01        23677.0   
7   E09000008                   croydon  1999-12-01        19563.0   
8   E09000009                    ealing  1999-12-01        20580.0   
9   E09000010                   enfield  1999-12-01        19289.0   
10  E09000011                 greenwich  1999-12-01        21236.0   
11  E09000012                   hackney  1999-12-01        23249.0   
12  E09000013    hammersmith and fulham  1999-12-01        25000.0   
13  E09000014                  haringey  1999-12-01        18783.0   
14  E09000015                    harrow  1999-12-01        20596.0   
15  E09000016                  havering  1999-12-01        17165.0   
16  E09000017                hillingdon  1999-12-01        24002.0   
17  E09000018                  hounslow  1999-12-01        20155.0   
18  E09000019                 islington  1999-12-01        25113.0   
19  E09000020    kensington and chelsea  1999-12-01        20646.0   
20  E09000021      kingston upon thames  1999-12-01        19302.0   
21  E09000022                   lambeth  1999-12-01        23151.0   
22  E09000023                  lewisham  1999-12-01        20580.0   
23  E09000024                    merton  1999-12-01        18962.0   
24  E09000025                    newham  1999-12-01        18862.0   
25  E09000026                 redbridge  1999-12-01        19580.0   
26  E09000027      richmond upon thames  1999-12-01        22321.0   
27  E09000028                 southwark  1999-12-01        22784.0   
28  E09000029                    sutton  1999-12-01        19582.0   
29  E09000030             tower hamlets  1999-12-01        26376.0   
30  E09000031            waltham forest  1999-12-01        18547.0   
31  E09000032                wandsworth  1999-12-01        21321.0   
32  E09000033               westminster  1999-12-01        24447.0   
33  E12000001                north east  1999-12-01        16282.0   
34  E12000002                north west  1999-12-01        16977.0   
35  E12000003  yorkshire and the humber  1999-12-01        16527.0   
36  E12000004             east midlands  1999-12-01        16392.0   
37  E12000005             west midlands  1999-12-01        17000.0   
38  E12000006                      east  1999-12-01        18000.0   
39  E12000007                    london  1999-12-01        22487.0   
40  E12000008                south east  1999-12-01        18737.0   
41  E12000009                south west  1999-12-01        16727.0   
42  E13000001              inner london  1999-12-01            NaN   
43  E13000002              outer london  1999-12-01            NaN   
44  E92000001                   england  1999-12-01        17939.0   
45  K02000001            united kingdom  1999-12-01        17803.0   
46  K03000001             great britain  1999-12-01        17866.0   
47  K04000001         england and wales  1999-12-01        17974.0   
48  N92000002          northern ireland  1999-12-01        15798.0   
49  S92000003                  scotland  1999-12-01        16914.0   
50  W92000004                     wales  1999-12-01        16457.0   

    life_satisfaction mean_salary recycling_pct  population_size  \
0                 NaN       48922             0           6581.0   
1                 NaN       23620             3         162444.0   
2                 NaN       23128             8         313469.0   
3                 NaN       21386            18         217458.0   
4                 NaN       

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [66]:
url= "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
london_data = pd.read_csv(url)
import numpy as np
print (type(london_data))
print( london_data.info())


def normalise_income(df):
    if (df['median_salary'] <= 27441):
        return 1
    elif (df['median_salary'] <= 30932):
        return 2
    else:
        return 3

london_data["median_salary"] = london_data.apply(normalise_income, axis=1)
london_data

  

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB
None


code                  area        date  median_salary  \
0     E09000001        city of london  1999-12-01              3   
1     E09000002  barking and dagenham  1999-12-01              1   
2     E09000003                barnet  1999-12-01              1   
3     E09000004                bexley  1999-12-01              1   
4     E09000005                 brent  1999-12-01              1   
...         ...                   ...         ...            ...   
1066  K03000001         great britain  2019-12-01              2   
1067  K04000001     england and wales  2019-12-01              2   
1068  N92000002      northern ireland  2019-12-01              1   
1069  S92000003              scotland  2019-12-01              2   
1070  W92000004                 wales  2019-12-01              2   

      life_satisfaction mean_salary recycling_pct  population_size  \
0                   NaN       48922             0           6581.0   
1                   NaN       23620             3         162444.0   
2                   NaN       23128             8         313469.0   
3                   NaN       21386            18         217458.0   
4                   NaN       20911             6         260317.0   
...                 ...         ...           ...              ...   
1066                NaN       37603           NaN              NaN   
1067                NaN       37865           NaN              NaN   
1068                NaN       32083           NaN              NaN   
1069                NaN       34916           NaN              NaN   
1070                NaN       31251           NaN              NaN   

      number_of_jobs  area_size  no_of_houses  borough_flag  
0                NaN        NaN           NaN             1  
1                NaN        NaN           NaN             1  
2                NaN        NaN           NaN             1  
3                NaN        NaN           NaN             1  
4                NaN        NaN           NaN             1  
...              ...        ...           ...           ...  
1066             NaN        NaN           NaN             0  
1067             NaN        NaN           NaN             0  
1068             NaN        NaN           NaN             0  
1069             NaN        NaN           NaN             0  
1070             NaN        NaN           NaN             0  

[1071 rows x 12 columns]

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [79]:
url= "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
london_data = pd.read_csv(url)

london_data["number_of_jobs"].describe()


def normalise_jobs(df):
    if (df['number_of_jobs'] <=london_data["number_of_jobs"].mean()):
        return 1
    elif (df['number_of_jobs'] <= london_data["number_of_jobs"].quantile(q=0.75)):
        return 2
    else:
        return 3

london_data[""] = london_data.apply(normalise_jobs, axis=1)
london_data


# print(london_data.describe("number_of_jobs"))

code                  area        date  median_salary  \
0     E09000001        city of london  1999-12-01        33020.0   
1     E09000002  barking and dagenham  1999-12-01        21480.0   
2     E09000003                barnet  1999-12-01        19568.0   
3     E09000004                bexley  1999-12-01        18621.0   
4     E09000005                 brent  1999-12-01        18532.0   
...         ...                   ...         ...            ...   
1066  K03000001         great britain  2019-12-01        30446.0   
1067  K04000001     england and wales  2019-12-01        30500.0   
1068  N92000002      northern ireland  2019-12-01        27434.0   
1069  S92000003              scotland  2019-12-01        30000.0   
1070  W92000004                 wales  2019-12-01        27500.0   

      life_satisfaction mean_salary recycling_pct  population_size  \
0                   NaN       48922             0           6581.0   
1                   NaN       23620             3         162444.0   
2                   NaN       23128             8         313469.0   
3                   NaN       21386            18         217458.0   
4                   NaN       20911             6         260317.0   
...                 ...         ...           ...              ...   
1066                NaN       37603           NaN              NaN   
1067                NaN       37865           NaN              NaN   
1068                NaN       32083           NaN              NaN   
1069                NaN       34916           NaN              NaN   
1070                NaN       31251           NaN              NaN   

      number_of_jobs  area_size  no_of_houses  borough_flag     
0                NaN        NaN           NaN             1  3  
1                NaN        NaN           NaN             1  3  
2                NaN        NaN           NaN             1  3  
3                NaN        NaN           NaN             1  3  
4                NaN        NaN           NaN             1  3  
...              ...        ...           ...           ... ..  
1066             NaN        NaN           NaN             0  3  
1067             NaN        NaN           NaN             0  3  
1068             NaN        NaN           NaN             0  3  
1069             NaN        NaN           NaN             0  3  
1070             NaN        NaN           NaN             0  3  

[1071 rows x 13 columns]

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1 

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer: 

## What caused you the most difficulty?

Your answer: 